scio-spanner/src/it/scala/com/spotify/scio/spanner/SpannerIOIT.scala

```diff
@@ -0,0 +1,148 @@
+    .withInstanceId("spanner-it")
+
+  private val adminClient = Spanner.adminClient(projectId)
+  private val dbClient = Spanner.databaseClient(config)
+
+  private final case class FakeSpannerData(asMutations: Seq[Mutation], asStructs: Seq[Struct])
+  private def fakeData(tableName: String) = FakeSpannerData(
+    Seq(
+      Mutation
+        .newInsertBuilder(tableName)
+        .set("Key").to(1L).set("Value").to("foo")
+        .build(),
+      Mutation
+        .newInsertBuilder(tableName)
+        .set("Key").to(2L).set("Value").to("bar")
+        .build()
+    ),
+    Seq(
+      Struct.newBuilder()
+        .set("Key").to(1L).set("Value").to("foo")
+        .build(),
+      Struct.newBuilder()
+        .set("Key").to(2L).set("Value").to("bar")
+        .build())
+  )
+}
+
+class SpannerIOIT extends FlatSpec with Matchers with BeforeAndAfterAll {
+  import SpannerIOIT._
+  implicit val ec: ExecutionContext = ExecutionContext.global
+
+  override def beforeAll(): Unit = {
+    adminClient
+      .createDatabase(
+        config.getInstanceId.get(),
+        config.getDatabaseId.get(),
+        List(
+          "CREATE TABLE read_query_test ( Key INT64, Value STRING(MAX) ) PRIMARY KEY (Key)",
+          "CREATE TABLE read_table_test ( Key INT64, Value STRING(MAX) ) PRIMARY KEY (Key)",
+          "CREATE TABLE write_test ( Key INT64, Value STRING(MAX) ) PRIMARY KEY (Key)"
+        ).asJava)
+      .waitFor()
+  }
+
+  override def afterAll(): Unit = {
+    adminClient.dropDatabase(config.getInstanceId.get(), config.getDatabaseId.get())
+  }
+
+  private class PopulatedSpannerTable(val tableName: String) {
+    val spannerRows: FakeSpannerData = fakeData(tableName)
+    dbClient.write(spannerRows.asMutations.asJava)
+  }
+
+  private class ReadableSpannerTable(val tableName: String) {
+    val writeData: FakeSpannerData = fakeData(tableName)
+
+    lazy val readOperationResults: List[Struct] = {
+      val txn = dbClient
+        .readOnlyTransaction()
+        .read(tableName, KeySet.all(), Seq("Key", "Value").asJava)
+
+      List(
+        { txn.next(); txn.getCurrentRowAsStruct },
+        { txn.next(); txn.getCurrentRowAsStruct }
+      )
+    }
+  }
+
+  "SpannerIO" should "perform writes" in new ReadableSpannerTable("write_test") {
+    private val sc = ScioContext(options)
+
+    SpannerWrite(config)
+      .writeWithContext(sc.parallelize(writeData.asMutations), SpannerWrite.WriteParam())
+
+    sc.close()
+
+    readOperationResults should contain theSameElementsAs writeData.asStructs
+  }
+
+  it should "perform reads from table" in new PopulatedSpannerTable("read_table_test") {
+    private val sc = ScioContext(options)
+
+    private val read = SpannerRead(config).readWithContext(
+      sc,
+      SpannerRead.ReadParam(
+        readMethod = SpannerRead.FromTable(tableName, Seq("Key", "Value")),
+        withTransaction = true,
+        withBatching = true
+      )
+    ).materialize.map(_.value.toList)
+
+    sc.close()
```

Comment:
@ClaireMcGinty 🎨  we could potentially re-use somethings here ```
sc
  .close()
  .waitUntilDone()

read.map(_ should contain theSameElementsAs spannerRows.asStructs)
```

Reply:
done!